File: homework/HW9/PythonVersion/LCAofTwoNodes.py
import TreeNode     #Import TreeNode.py

tree = TreeNode     #Create variable to use TreeNode.py

class LCAofTwoNodes:

    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if root is None or p is None or q is None:
            return None
        if root.val > p.val and root.val > q.val:
            return  self.lowestCommonAncestor(root.left, p , q)
        elif root.val < p.val and root.val < q.val:
            return self.lowestCommonAncestor(root.right, p , q)
        else:
            return root


#Create given tree
root = tree.TreeNode(4) 
root.left = tree.TreeNode(3)
root.left.left = tree.TreeNode(1)
root.right = tree.TreeNode(8)
root.right.left = tree.TreeNode(5)
root.right.right = tree.TreeNode(9)

#Get test variables
p = root.right.left
q = root.right

#Variable to use lca function
test = LCAofTwoNodes()


ans = test.lowestCommonAncestor(root,p,q)

print("LCA (" + str(p.val) + ", " + str(q.val) + "): " + str(ans.val))

File: homework/HW9/JavaVersion/LCAofTwoNodes.java
package homework.HW9.JavaVersion;


public class LCAofTwoNodes{

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if(root == null){
            return null;
        }
        if(root.val > p.val && root.val > q.val){
            return lowestCommonAncestor(root.left, p, q);
        }
        else if(root.val < p.val && root.val < q.val){
            return lowestCommonAncestor(root.right, p, q);
        }
        else{
            return root;
        }

    }
    public static void main(String[] args) {
        LCAofTwoNodes test = new LCAofTwoNodes();

        TreeNode root = new TreeNode(4);
        root.left = new TreeNode(3);
        root.left.left = new TreeNode(1);
        root.right = new TreeNode(8);
        root.right.left = new TreeNode(5);
        root.right.right = new TreeNode(9);

        /* Given tree:
         *     4
         *    / \ 
         *   3   8
         *  /   / \
         * 1   5   9
         */

        TreeNode p = root.right;        //Test Variable
        TreeNode q = root.right.left;   //Test Variable
        

        TreeNode ans = test.lowestCommonAncestor(root, p, q);
        System.out.println("LCA (" + p.val + ","+ q.val + "): " + ans.val);

    }
    
}

